pkg/collector/python/helpers.go

```diff
@@ -122,21 +121,56 @@ func (sl *stickyLock) unlock() {
 	runtime.UnlockOSThread()
 }
 
-// cStringArrayToSlice returns a slice with the contents of the char **tags (the function will not free 'array').
-func cStringArrayToSlice(array **C.char) []string {
-	if array != nil {
-		res := []string{}
-		for i := 0; ; i++ {
-			// Work around go vet raising issue about unsafe pointer
-			strPtr := C.getStringAddr(array, C.uint(i))
-			if strPtr == nil {
-				return res
-			}
-			str := C.GoString(strPtr)
-			res = append(res, str)
-		}
+// cStringArrayToSlice converts an array of C strings to a slice of Go strings.
+// The function will not free the memory of the C strings.
+func cStringArrayToSlice(a **C.char) []string {
+	if a == nil {
+		return nil
+	}
+
+	var length int
+	forEachCString(a, func(_ *C.char) {
+		length++
+	})
+	res := make([]string, 0, length)
+	si, release := acquireInterner()
+	defer release()
+	forEachCString(a, func(s *C.char) {
+		bytes := unsafe.Slice((*byte)(unsafe.Pointer(s)), cstrlen(s))
+		res = append(res, si.intern(bytes))
+	})
+	return res
+}
+
+// cstrlen returns the length of a null-terminated C string. It's an alternative
+// to calling C.strlen, which avoids the overhead of doing a cgo call.
```

Comment:
Do you have an idea of the impact in the current benchmark of avoiding the cgo call for strlen?

Reply:
I just did a standalone benchmark of calling `C.strlen` (see diff below) against the Go implementation. Seems like doing it in Go is about 3x faster than taking the hit of the cgo call.

This is actually a bigger benefit than I would have expected. My mental model is that cgo calls carry an overhead of [maybe 40ns](https://shane.ai/posts/cgo-performance-in-go1.21/), but here we seem to see an order of magnitude more overhead.

I'm not immediately sure what's causing this, but it could be an arm64 vs amd64 issue, or related to using docker on mac, or maybe a dynamic linking issue with libc. A quick look at the profile showed [_cgo_topofstack](https://github.com/golang/go/blob/fa72f3e034fdabc5922ac019281f53ea0a8328cf/src/runtime/asm_arm64.s/#L1175-L1190) using a lot of time which is surprising, considering that it's just a few instructions 🤷.

Anyway, I even if we saw 40ns per cgo call, it would still yield 20% savings on this benchmark, so I think doing `strlen` in Go is definitely worth it.

```
$ benchstat c_strlen.txt go_strlen.txt 
goos: linux
goarch: arm64
pkg: github.com/DataDog/datadog-agent/pkg/collector/python
                        │ c_strlen.txt │            go_strlen.txt             │
                        │    sec/op    │    sec/op     vs base                │
_cStringArrayToSlice-10    692.2n ± 6%   212.9n ± 13%  -69.24% (p=0.000 n=10)

                        │ c_strlen.txt │         go_strlen.txt          │
                        │     B/op     │    B/op     vs base            │
_cStringArrayToSlice-10     160.0 ± 0%   160.0 ± 0%  ~ (p=1.000 n=10) ¹
¹ all samples are equal

                        │ c_strlen.txt │         go_strlen.txt          │
                        │  allocs/op   │ allocs/op   vs base            │
_cStringArrayToSlice-10     1.000 ± 0%   1.000 ± 0%  ~ (p=1.000 n=10) ¹
```

```diff
diff --git i/pkg/collector/python/helpers.go w/pkg/collector/python/helpers.go
index 445f2de2ee..865c19712c 100644
--- i/pkg/collector/python/helpers.go
+++ w/pkg/collector/python/helpers.go
@@ -136,7 +136,7 @@ func cStringArrayToSlice(a **C.char) []string {
        si, release := acquireInterner()
        defer release()
        forEachCString(a, func(s *C.char) {
-               bytes := unsafe.Slice((*byte)(unsafe.Pointer(s)), cstrlen(s))
+               bytes := unsafe.Slice((*byte)(unsafe.Pointer(s)), C.strlen(s))
                res = append(res, si.intern(bytes))
        })
        return res
```

![2024-03-11 python test cpu at 20 37 13@2x](https://github.com/DataDog/datadog-agent/assets/15000/136a72c2-b9a7-47e4-965a-2e2f1e570cc5)